### ISSUE TITLE
PutLink definition

### ISSUE THREAD

**Issue by ngeiswei** [opencog/atomspace#1552]
It's not clear to me what
```
Put
  Lambda
    Variable "$X"
    Variable "$X"
  List
    Concept "A"
    Variable "$Y"
```
should reduce to.

First, the List seems to indicate that there are 2 arguments but the Lambda has only 1. Second, should the result be a Lambda or a List?

Currently, executing that PutLink outputs
```
List
  Concept "A"
  Variable "$Y"
```
which I think is incorrect, it should either

1. raise an exception because of argument number mismatch
2. or, if it allows it, wrap a Lambda since the argument contain a variable, to obtain
```
Lambda
  List
    Concept "A"
    Variable "$X"
```

BTW, the reason I say that it should be wrapped by a Lambda is just to be consistent with the current behavior, such as `(Put (Lambda (Inheritance X Y)) (List (Concept "A") Z)))`  reduces to `(Lambda (Inheritance (Concept "A") Z))`, which is different than what lambda calculus would dictate (as it would remove the abstraction after a beta-reduction, regardless of whether the arguments have variables of not). I don't have a problem not following exactly lambda calculus BTW, just pointing this out.

Also, I have no problem raising an exception is such case, i.e. I don't currently need to support that, but I was just wondering. Meanwhile, I think I'll just raise an exception.

**Comment by ngeiswei** [comment]
Based on the following comment https://github.com/opencog/atomspace/blob/master/opencog/atoms/core/PutLink.cc#L188 I'm gonna raise an exception of mismatched argument numbers.

**Comment by linas** [comment]
> First, the List seems to indicate that there are 2 arguments but the Lambda has only 1. 

The original design tried to count the number of variables, and give that precedence.  Since the lambda has one variable, then the List is what should be pasted in there.  There was a vague idea that this would allows List to sometimes be treated as ordinary atoms.  But, yes, I agree that overall, this construction is ambiguous, and ripe for confusion and abuse.

**Comment by linas** [comment]
> Second, should the result be a Lambda or a List?

Lets pretend it was this:
```
Put
  Lambda
    Variable "$X"
    Variable "$X"
  FooBarLink
    Concept "A"
    Variable "$Y"
```
Now, since $Y is free in FooBar, it seems like it should also be free in the result. The result should be 
```
  FooBarLink
    Concept "A"
    Variable "$Y"
```
and the put did not cause $Y to be bound, when it wasn't before.

**Comment by linas** [comment]
> (Put (Lambda (Inheritance

Interesting example.  There, initially, the $Y was bound.  The logic there is that, by replacing it with $Z it should not become unbound -- that, instead, it should be treated as a variable renaming.  That seems like the right thing to do.  Yes, its not lambda calculus.  I don't know if that's good, or if that's bad.  I prefer to leave it bound, but that is purely a gut-instinct reasoning, and not a strong rational argument that I can articulate.

**Comment by ngeiswei** [comment]
Ah, OK, so I think I had misunderstood the following equivalence
```
List X Y Y Z
```
and
```
(List (Lambda (VariableList X Y Z) (List X Y Y Z)))
```
as presented in https://github.com/opencog/atomspace/issues/1490#issuecomment-353244646

and generalize that for any link type.

I'm not sure I can make sense of this equivalence either, but it is very convenient so...

**Comment by linas** [comment]
Yes, I agree its confusing. No, they are not a formal equivalence; instead, they just "accidentally" give the same results in many, but not all situations.  First, replace `List` by `Foobar`  Then `Foobar X Y Y Z` is a link with three free variables.

By contrast, `(Lambda (VariableList X Y Z) (Foobar X Y Y Z)))` is a scoped link with three bound variables.

Its... well, for right now, its probably not worth it to write down all the various possible variations in which these can be pasted around ... unless you want me to.  In most variations, they give the same results; in a few, they do not.

**Comment by ngeiswei** [comment]
It's OK for now.
